Core/include/Acts/AmbiguityResolution/GreedyAmbiguityResolution.ipp

```diff
@@ -0,0 +1,71 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/AmbiguityResolution/GreedyAmbiguityResolution.hpp"
+
+#include <unordered_map>
+
+namespace Acts {
+
+template <typename track_container_t, typename traj_t,
+          template <typename> class holder_t>
+void GreedyAmbiguityResolution::computeInitialState(
+    const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+    State& state) const {
+  std::unordered_map<SourceLink, std::size_t, decltype(m_cfg.sourceLinkHash),
+                     decltype(m_cfg.sourceLinkEquality)>
```

Comment:
I was wondering whether the delegates would add a noticeable overhead, and whether that could be avoided by providing the missing functionality via an adapter (extra template parameter)? But maybe my concerns are not justified.

Reply:
I am also not very happy about this delegate. IMO the source link should provide you with the functionality of checking if two are equal or not.

I guess having the hash and equality passed via a template would remove one level of indirection and allows for more compiler optimization. I can try and see if this also makes the code a bit nicer

Reply:
If the source link did that we'd have to use a similar type erasure mechanism as delegate itself. Delegate should have very little overhead for function calls, it's basically a function pointer call that can be inclined even. So I don't think this is something we need to be worried about.